### ISSUE TITLE
Add event API to allow low-level events (e.g. input)

### ISSUE THREAD

**Issue by beoran** [hajimehoshi/ebiten#926]
As per our discussion on Slack I'd like to propose that Ebiten exposes the low level events it receives from the underlying platform (key presses, joystick, etc) in a separate package "event". 

Arguments in favor: 

1) It's more efficient than converting the state based API.
2) It's easier to use for certain applications, such as GUI's.
3) In some cases it's more correct.

The API could be something like it is in Allegro, https://liballeg.org/a5docs/trunk/events.html,  but without the C-ism like unions and an event queue, which could be replaced simply with a Go channel.

**Comment by hajimehoshi** [comment]
Thanks! I agree to introduce a new package for lower level input. The existing input APIs would rely on the new 'event' package.

Note that gomobile has an event system by channels: https://godoc.org/golang.org/x/mobile/app

I'll reply more details about your idea later.

**Comment by hajimehoshi** [comment]
My assumption is that all event, not only inputs but also e.g., updating frames, should be in the same event loop, so the event package would treat all such events. (and the existing APIs would be the wrapper).  Do you think it'd make sense?

**Comment by beoran** [comment]
Yes, that makes sense. All events should be handled in the same loop. In Allegro we also have things like timer events, or events when the screen becomes not drawable, etc. The existing APi is the wrapper, and if you like, you can also request that the events are not sent only to the wrapper but also directly to your own channel.

Yes, the design of the event system could probably look like it is in x/mobile/app , although the events would be a bit more high level (wrapping the platform specific events).  In that library there are events for lifecycle, mouse, paint, size, touch. We will also need events for joystick, keyboard, window. Perhaps timers are not needed as Go already provides these.

**Comment by hajimehoshi** [comment]
Thanks. We're now on the same page :) Then, what we should do will be:

1) Rewrite driver.UI, driver.UIContext, driver.Input, (e.g., https://github.com/hajimehoshi/ebiten/blob/master/internal/driver/ui.go) event-based
2) Rewite all the driver implementation
3) Create the new package github.com/hajimehoshi/ebiten/event to use the new driver API
4) Rewrite the existing package github.com/hajimehoshi/ebiten to use the event package

Could you write a doc about your suggestion of the new API in 'event' package?

**Comment by beoran** [comment]
It's easier for me to write some code than a design document, so I wrote this documented code in the hope it will be clear enough to show what I mean.  Feel free to make suggestions, the file is here:

https://github.com/beoran/ebiten/blob/beoran-event/event/event.go

**Comment by hajimehoshi** [comment]
Hi, thank you for your suggestion! I looked through and would like to point out:

* Is there a reference you used to define event types?
  * Web API might be the most portable.
* Let's add game-updating event (`KindGameUpdate`) that calls the registered function to update the game state
  * `DisplayPaint` is similar but Ebiten can call the function multiple times for one display update.
* Let's use a standard `time.Time` instead of `Timestamp`
* Rename `Code` -> `KeyCode`, `Modifier` -> `KeyModifier`, and so on
* Rename `Joystcick` -> `Gamepad`
* Rename `Axis` -> `GamepadAxis`, `Button` -> `GamepadButton`, and so on
* Display events will be necessary but sound hard to define so far. Could we skip defining them for now and consider them later?
  * I prefer the word 'View' to 'Display', but this is controversial. 'Display' might be confusing with physical monitors.
  * Ebiten doesn't have 'orientation' change events.
  * It is now chaos how Ebiten apps treat screen-size changes.
    * Apps on desktops and browsers assume that the screen size never changes unless `SetScreenSize` is called.
    * Apps on mobiles assume the same thing, and if the outer view (a GL surface)'s size is changed, the scale is adjusted automatically (when `ebitenmobile` command is used).
* Application lifetime is not determined.Could we skip defining application events for now and consider them later?
  *  Mobiles and browsers don't have 'destroy'. 
* I don't think we have to define a type for a channel.

**Comment by hajimehoshi** [comment]
> * Let's add game-updating event (KindGameUpdate) that calls the registered function to update the game state
>   * DisplayPaint is similar but Ebiten can call the function multiple times for one display update.

On second thought, as `event` should be the lower-level layer, `DisplayUpdate` (or other names like `FrameUpdate`) sounds better and the user of `event` package should calculate how many times the update function is called 🤔

**Comment by beoran** [comment]
I have updated the file based on your suggestions:
https://github.com/beoran/ebiten/blob/beoran-event/event/event.go

To answer your questions:
1. I used Allegro as the reference since it has all useful events I can think of.
2. I made an event type KindViewUpdate, that could be used as you suggest.
3. I used time.Time everywhere.
4. I added prefixes for the struct member types everywhere, for Key, Mouse, Touch, Gamepad.
5. I used Gamepad everywhere.
6. Did that too as you suggested.
7. I renamed Display to View, and removed the KindViewOrientation event, but kept the KindViewResize event since you will need that to adjust the scale automatically on a higher level.  We can define other View events as the need arises.
8. I dropped all Application events, we can define them later if needed.
9. I removed the channel type
10. I added a TouchID, and renamed Which to GamepadID as well, for consistency.

Hope you will like my changes and answers. :)

**Comment by hajimehoshi** [comment]
Thank you!

> I renamed Display to View, and removed the KindViewOrientation event, but kept the KindViewResize event since you will need that to adjust the scale automatically on a higher level. We can define other View events as the need arises.

OK I agreed.

* Can we have `Basic` struct without `Any` struct?
```

**Comment by hajimehoshi** [comment]
Thank you!

> I renamed Display to View, and removed the KindViewOrientation event, but kept the KindViewResize event since you will need that to adjust the scale automatically on a higher level. We can define other View events as the need arises.

OK I agreed.

* What is the `Source` object? What kind of object is there?
* Do we really need `Time` attribute at `Basic`? Wouldn't the time when receiving enough?
* Do we really need `Event` interface? I think we can eliminate `Any` struct from `Basic` struct. `Any` doesn't sound idiomatic in Go, but perhaps I might not know the idiomatic.
* We don't use Z/W positions of mice.
* When do we use `NewChannel`? I thought the loop would be like:

```
for := range  {
  
}
```

**Comment by hajimehoshi** [comment]
Thank you!

> I renamed Display to View, and removed the KindViewOrientation event, but kept the KindViewResize event since you will need that to adjust the scale automatically on a higher level. We can define other View events as the need arises.

OK I agreed.

* What is the `Source` object? What kind of object is there?
* Do we really need `Time` attribute at `Basic`? Wouldn't the time when receiving enough?
* Do we really need `Event` interface? I think we can eliminate `Any` struct from `Basic` struct. `Any` doesn't sound idiomatic in Go, but perhaps I might not know the idiomatic.
* We don't use Z/W positions of mice.
* When do we use `NewChannel`? I thought the loop would be like:

```
app.Main(func(eventCh chan Event) {
for e := range eventCh {
  if e.Kind() == KindMouseEnter {
    // ...
  }
}
```

**Comment by beoran** [comment]
I updated the file again.

To answer your questions:

1. Source is for the event's low level source, i. e. the driver. I included it because my experience from Allegro shows that it can be useful to have a reference to that.
2. Time is the time when the event was /emitted/. It's possible that some events are too slow in and will need to be discarded. I copied this from Allegro.
3. Event needs to be an interface for type safety and to allow users to reuse the interface to send their own events. This is very useful for GUI systems. But it's true we don't need the Any sub struct so I solved this in a different way.
4. I dropped Z and W, although maybe mouse wheel would be useful later.
5. You are right, we will probably not need this function, so I dropped it.

**Comment by hajimehoshi** [comment]
Thank you!

> Source is for the event's low level source, i. e. the driver. I included it because my experience from Allegro shows that it can be useful to have a reference to that.

I understood you referred Allegro, but let's remove this until we find an actual use case. I don't want to expose a driver things of Ebiten.

> Time is the time when the event was /emitted/. It's possible that some events are too slow in and will need to be discarded. I copied this from Allegro.

Ditto.

> Event needs to be an interface for type safety and to allow users to reuse the interface to send their own events. This is very useful for GUI systems. But it's true we don't need the Any sub struct so I solved this in a different way.

Let's revisit after we solve the above items.

**Comment by beoran** [comment]
OK, I updated the file again. I left out Source and Time, and renamed the method Kind in the event interface to EventKind, so Basic can simply have a Kind member.  I also added a TouchPressure member to the Touch event, and removed DZ and DW. 

I am still convinced that Event needs to be an interface as there are no C-like unions in Go. The best way in Go to simulate unions is to use an interface.

**Comment by hajimehoshi** [comment]
Thank you for fixing the proposal again and again!

* We no longer need `Source` interface.
* Do we need Z/W position for mouse?

> I am still convinced that Event needs to be an interface as there are no C-like unions in Go. The best way in Go to simulate unions is to use an interface.

I think there is still room for discussion. As the interface now has only 'EventKind', if we had another way to determine the kind of an event, we would no longer need the interface.

Could we determine the event type by type assertion instead of a kind value? If you need a custom event, you would need a new kind value that is never duplicated with existing value. Even if you could create a new kind value that is not duplicated with other existing events, other people might add a new kind that can duplicate with yours... On the other hand, type assertion way can avoid this problem completely.

What do you think?

**Comment by beoran** [comment]
I made another more "radical" change this time, so type switches or type assertions wi work. I still kept the Event interface with a marker method IsEvent, because I dislike the empty interface. With the Event interface it is hard to accidentally pass the wrong object, and easy to add new members later if needed.

I also simplified even more by dropping most integer types, that didn't do much. I renamed all D methods to Delta, that is clearer.

I did add a Wheel and DeltaWheel for the mouse, because they are useful for scrolling. 

Hopefully I am getting closer to to make this idea to your liking.

**Comment by hajimehoshi** [comment]
Thanks! I think you agreed that type-assertion way is better :-)

* I was wondering why we need a marker method. As far as I know, there is only-one use case of a maker method: https://golang.org/pkg/runtime/#Error , but this is used to distinguish from regular errors. When sending a custom event, would we really make a mistake as you said? Or, even if we send an arbitrary object to the event emitter, what would be the problem?
* What about ending with `Event` if a struct is an emitted event struct? There are structs that are not emitted there.
* What about integrating touch-related events to `TouchEvent` and use an enum value (e.g., `TouchType`)?

> Hopefully I am getting closer to to make this idea to your liking.

Yeah, I think we are getting closer to the agreement :-)

**Comment by beoran** [comment]
Another update!

1. Well the main problem would be forward compatibility in case we decide we need some top-level methods on all events after all. But yes, the marker method is not strictly needed, so I'll make `type Event interface {}`. this still makes type checking possible. I also dropped the Basic struct.
   
2. That would create too much of a stutter, since other packages will use the events as event.KeyUp, event.KeyDown, etc. It would be a hassle to have to type event.KeyUpEvent or worse, event.EventKeyUp. However it's true that some structs are just auxilliary. So I made them private and added constructors, which seems to give a better user experience as well.   

3. What you suggest is actually contrary to what we are trying to do in eliminating the EvenrKind. Now, with every event it's own struct, we can handle all events with a single type switch ot a series of types assertions. If TouchKind becomes separated out then that will needs a separate second select statement to handle touch events only. Also it would be confusing between the different events. We shoud either have an Eventkind or not, but not a mix of both.   

I added a small try out program as well, here:
https://github.com/beoran/ebiten/blob/beoran-event/event/tryevent/tryevent.go

**Comment by hajimehoshi** [comment]
>  It would be a hassle to have to type event.KeyUpEvent or worse

I tend to agree, but, `New*` functions sound verbose and I'd prefer plain structs for events... What about 'E' (e.g., `event.KeyDownE`)?

> What you suggest is actually contrary to what we are trying to do in eliminating the EvenrKind. 
> ...
> If TouchKind becomes separated out then that will needs a separate second select statement to handle touch events only. Also it would be confusing between the different events. We shoud either have an Eventkind or not, but not a mix of both.

I meant to eliminate entire EventKind, but I'm fine having enums to distinguish details for one group of events (key up or key down or key press...). It's because my motivation to eliminate kind was to make it easy to add a custom event, and adding auxiliary enums should not prevent it.

However, your idea makes sense. Mixing might be confusing. I'd like to see plain-structs way as I suggested anyway.

> https://github.com/beoran/ebiten/blob/beoran-event/event/tryevent/tryevent.go

Nice! That sounds pretty good!

**Comment by beoran** [comment]
OK, I think I understand, you really like to keep things /simple/. :-)

So, I dropped the New functions, and removed all sub structs. Now, It's very easy to make an event using, for example,  `event.KeyUp{64, 0}`. A E suffix is even less readable than an Event suffix, but since all structs are now flat and there are no structs that are not events anymore, I think we don't need any suffix anymore. Look at the tryevent.go to see how nicely it looks now.

Mixing events and kinds for sub events is not only somewhat confusing, also less efficient, because you need two checks in stead of one. In the I have a handler for all possible events, 19 in total, and I think it looks fine. Normally we will branch out to different handler functions in a GUI anyway.

Thanks for your appreciation.

**Comment by hajimehoshi** [comment]
I'm almost satisfied with your suggestion https://github.com/beoran/ebiten/blob/beoran-event/event/event.go. Thank you! That is very beautiful and, yes, very simple :-). I still have small requests, but l'll leave them until an actual PR comes.

So, next, let's tackle actual action items as I mentioned at https://github.com/hajimehoshi/ebiten/issues/926#issuecomment-523803963 . 3 can be implemented as a mock without 1. and 2., but what we could do would be limited. @beoran , do you have a preference what you want to do? If not, could you make a PR for 3.?

**Comment by beoran** [comment]
I made #930 with a PR, for only 3. I tried to follow the guide lines but I ill need your guidance.
Once this is OK, we can work on action points 1., 2. and 4. in that order, I guess.

**Comment by hajimehoshi** [comment]
Action items I come up with:

  1. Add `event.Key` definition and make `ebiten.Key` an alias to `event.Key` (@hajimehoshi will do)
  2. Add `event.ModifierKey` (or `Modifier`?)
  3. Add a new function to `driver.UI` to return a channel that emits events.
  4. Implement the function on GLFW.
  5. Make an example that uses `event` directly.
  6. Implement the function on JS.
  7. Implement the function on mobiles.
  8. Reimplmement `ebiten` package to use `event` package.

Now I'm doing 1. Stay tuned!

**Comment by beoran** [comment]
Yes, that seems like a good plan. Although, maybe 10. needs to be done between 4. and 5. If you feel you need help with any of those, please ask and I'll do what I can.

**Comment by hajimehoshi** [comment]
Yes. Especially gamepad part is relatively independent thing, so I'm interested in your plan about gamepads. I think you will refer Allegro?

**Comment by beoran** [comment]
Yes: https://gitlab.com/beoran/galago/tree/master.

I want to make a low lever replacement for GLFW, but I will make it completely modular, so you can ue only what you need, and I will start with the game pad part, as I wrote the gamepad/joystick and haptic drivers for Allegro5.

**Comment by hajimehoshi** [comment]
Finished 1. and 2. at https://github.com/hajimehoshi/ebiten/issues/926#issuecomment-526936431

**Comment by hajimehoshi** [comment]
There are three packages that we consider the relationships

  * `event`
  * `internal/driver`
  * driver implementations like `internal/uidriver/glfw` 

`internal/uidriver/glfw` depends on `internal/driver` for key definitions or other interface definitions. The question is whether `event` should depend on `internal/driver` or vice versa.

My current plan is to make `event` depend on `internal/driver`.

```go
package event

func Loop() chan interface{} { // Event interface should be used?
  return getAppropriateDriver().Loop()
}
```

```go
package driver

type UI interface {
   Loop() chan interface{}
}
```

The problem is that the `internal/driver` package cannot know event structs like `KeyboardKeyDown`. The driver side would need to implement counterparts of event structs, and event package converts this appropriately.

If `internal/driver` depends on `event`, as driver implementations depend on `internal/driver`, I'm not sure how `event` determines actual drivers.

@beoran Any thoughts?

**Comment by beoran** [comment]
Hmm, the dependency is a bit of a problem, yes. I think we should probably
split the event package into a internal/event package with the types and
constants, and an event package, with functions and aliases. Then the
dependency graph becomes internal/event -> internal/driver -> event and
also internal/event-> event.

Does that sound like a good idea?

Op di 3 sep. 2019 16:35 schreef Hajime Hoshi <notifications@github.com>:

**Comment by hajimehoshi** [comment]
So where would the event struct definition be? At least `event` should have them for the users, right?

**Comment by beoran** [comment]
Hmm, we could use aliases, but that would not be so readable. So no, my
first idea isn't that good.

When I think more, the problem is not the structs, but the constants such
as the key codes. We would like to map these to the key constants of the
glfw driver, but at the same time we want to use the events in the driver,
which would create a cyclical dependency, because the events use said key
constants.

To break this cycle, perhaps it's easier to duplicate the key constants.
Because you are generating them automatically anyway?



Op di 3 sep. 2019 19:02 schreef Hajime Hoshi <notifications@github.com>:

**Comment by hajimehoshi** [comment]
Key constants are already shared between `internal/driver` and `event`. They are generated automatically.

I think aliases sound the way to go. The original definition will be in `internal/driver` and `event` define them for their aliases. The problem is that godoc.org would not show the original definitions and this would make the users confused. 🤔

**Comment by beoran** [comment]
If we are going to use aliases, them we probably have to take the output of
go doc internal/event, and then paste that above the aliases as a comment
in event. The file event/event.go with aliases and comments could then be
generated, perhaps?

Op di 3 sep. 2019 19:38 schreef Hajime Hoshi <notifications@github.com>:

**Comment by hajimehoshi** [comment]
Generating sounds good. My current idea is generate struct definitions that are duplicated, and convert it like `event.Key` and `driver.Key`. It is easy to convert a struct to another struct as long as they have exported same members (https://play.golang.org/p/Mizflu9T3qc). Duplication might not be beautiful, but this should work.

**Comment by beoran** [comment]
Yes, I guess some duplication might be unavoidable if only for documentation or dependency reasons.

**Comment by hajimehoshi** [comment]
Now I'm making an auto generator for events. Stay tuned...

**Comment by beoran** [comment]
Good, I am still working slowly on a Linux input/joystick/Gamepad driver.

**Comment by beoran** [comment]
The approach of using a separate Go joystick driver for Linux may work, but perhaps it's easier and quicker for Ebiten if GLFW supported a callback function when the joystick changes. I read in their issue tracker something like this is desired but no one has done the effort to implement it. I guess that I could do this in stead.

**Comment by hajimehoshi** [comment]
As I finished works for 1.10, I'll restart this work.

**Comment by hajimehoshi** [comment]
> 1. Add event.Key definition and make ebiten.Key an alias to event.Key (@hajimehoshi will do)
> 2. Add event.ModifierKey (or Modifier?)
> 3. Add a new function to driver.UI to return a channel that emits events.
> 4. Implement the function on GLFW.
> 5. Make an example that uses event directly.
> 6. Implement the function on JS.
> 7. Implement the function on mobiles.
> 8. Reimplmement ebiten package to use event package.
> 9. Remove unnecessary functions from driver.UI and Input interface.
> 10. Improve GLFW driver to use an event-driven style

My first plan was to make `ebiten` package depend on `event` package. In this plan, theoretically users can develop apps only with `event` package without `ebiten` package. I thought this would be clean but the change would be huge. In the first place, what we needed was fined grained input detection, then creating a package only for events sounds overkill.

Instead, my suggestion is to create `Events` function that returns a channel in `ebiten` package, and `event` packages will include only event definitions.

The new action items are:

- [x] 1. Add event.Key definition and make ebiten.Key an alias to event.Key (@hajimehoshi will do)
- [x] 2. Add event.ModifierKey (or Modifier?)
- [ ] 3. Add a new function to driver.UI to return a channel that emits events.
- [ ] 4. Implement the function on GLFW.
- [ ] 5. (Updated!) Add `Events` function to `ebiten` package and create an example.
- [ ] 6. Implement the function on JS.
- [ ] 7. Implement the function on mobiles.
- [ ] 8. (Updated!) Reimplmement `ebiten` package to use `Events` function. (This is still unclear. Would it be possible to keep both the current input API implemented with `Events` and exposing the API `Events`?)
- [ ] 9. Remove unnecessary functions from driver.UI and Input interface.
- [ ] 10. Improve GLFW driver to use an event-driven style

@beoran What do you think?

**Comment by beoran** [comment]
Thanks for getting back top this! This plan looks fine, we can still move gradually to use events more in Ebiten later on.

The issue for the event callback for joysticks for GLFW is here: 
https://github.com/glfw/glfw/issues/601
I am having some problems with this GLFW issue, though, since joysticks and game pads in GLFW are actively polled when you read them. There are functions to poll the event state and call the callback functions but these also have to be changed quite thoroughly to also support joystick events.

**Comment by beoran** [comment]
I made some progress on GLFW joystick events, they work on Linux with my PR. As you know, I can only test on Linux, perhaps, if you could spare the effort, you could test my PR on other platforms as well?

**Comment by hajimehoshi** [comment]
> EDIT: one important thing to note is that Ebiten will have to call
glfwPollEvents() regularly (perhaps once every frame) to receive the events through the callbacks.

Ah good point. Thanks!

> I made some progress on GLFW joystick events, they work on Linux with my PR. As you know, I can only test on Linux, perhaps, if you could spare the effort, you could test my PR on other platforms as well?

Are you talking about https://github.com/glfw/glfw/pull/1590 ? Now Ebiten uses GLFW 3.2, and I am waiting for GLFW 3.3 binding for Go. Then before testing your PR, I need to move on GLFW 3.3...

**Comment by beoran** [comment]
If you need bindings, I'll contribute those too, I have made bindings for Allegro 5 to Go as well. But I'd prefer my PR for GLFW to be accepted first, so I don't have to do double work.

**Comment by hajimehoshi** [comment]
There is a PR for GLFW 3.3 https://github.com/go-gl/glfw/pull/256.

Even after this PR is landed, we need to wait for 3.4 including your PR, right?

**Comment by beoran** [comment]
Yes, likely so. Hmmm, updating GLFW might not be the fatstest way to get events for joystick input after all, who knows when they will make 3.4? At this rate, maybe in 2022? Hopefully not, but if they are slow in taking in my PR, I'll probably have to go back to my previous idea of programming a low level game library in Go myself to replace GLFW.

**Comment by beoran** [comment]
I recently discovered Fyne (https://github.com/fyne-io/fyne). This is a Go langage GUI library that also uses GLFW 3.2 and is fully event based. It seems they set up callbacks for mouse and keyboard events. https://github.com/fyne-io/fyne/blob/78cdfc5936493bfcc77a99e0a6e26dcf8ea42ad7/internal/driver/gl/window.go

The only thing missing really are joystick events. However, on Linux, there's no need to use GLFW for joysticks, you could just use my library in stead. On other platforms, I don't know... :p

**Comment by hajimehoshi** [comment]
OK, I'll take a look. What is the URL of your library? Is this in pure Go?

**Comment by beoran** [comment]
It's the same as above, look here for how to use it:
https://gitlab.com/beoran/galago/blob/master/os/linux/input/input_linux_test.go
Admittedly, I have to clean it up more, so your suggestions are welcome. 

This is pure Go, as I interact with the Linux kernel directly. That's the great thing of the  Linux kernel, it has a stable ABI, so, you can directly do system calls to it, without having to use any C at all.

**Comment by hajimehoshi** [comment]
Hmm, this works only on Linux, right?

**Comment by beoran** [comment]
Yes, right, I am focusing on Linux right now, other platforms are harder to
support for me right now.

Op di 3 dec. 2019 14:07 schreef Hajime Hoshi <notifications@github.com>:

**Comment by hajimehoshi** [comment]
Sorry but let me remove the milestone from this issue. This issue seems a very long-term work.

**Comment by beoran** [comment]
Yes, that's right. Only now someone on GLFW tried out my branch on Windows. My own messing around doesn't make much progress for now too. We'll have to wait for the situation of either GLFW or of myself to get better. :)

**Comment by hajimehoshi** [comment]
Merged into #1704
